### PR TITLE
Use pip-compile to help with consistent Python dependency resolution

### LINKED ISSUE #223: Reproducible builds
Starting this discussion with an initial proposal of how to support reproducible builds to freeze the ecosystem state leading up to a release.

# Problem
At the moment, we have not implemented a mechanism to pin version `X` of jax, `Y` of paxml, and `Z` of a transitive dependency of paxml in our containers; where X, Y, and Z can be versions on pypi, git-refs, or even a full distribution. Currently, we rely on nightly images to pin state, but these are not reproducible since running the same build script may install more recent libraries or transitive dependencies that break functionality.

This is an issue when we want to rebuild a container for a release, because we become cautious of changing build commands in previous layers (e.g., jax or pax) in fear that the build would pick up an unintentional change.

## Requirements
There are two things we need to freeze:
1. The version of a python package, e.g., pax, jax, transformer-engine
2. The distribution of a library, e.g., pax + (all of our patches)

## Proposal
Here is my initial proposal, which we can iterate on.

### Freezing python package versions
There are mature solutions to handle python package dependencies like pip-tools (pip-compile) and we can make use of those tools here. We can have:
* requirements-jax.in
* requirements-pax.in
* requirements-t5x.in

Where the jax images install `requirements-jax.in` and the t5x images installs `pip_compiled(requirements-jax.in, requirements-t5x.in)`.

### Freezing the state of the nightly
Previously, we froze the state of all upstream repos by building it in a docker image. In order to re-build that image, we will need to update our CI to do something like the following:

> (get git-sha of all upstream repos nightly) -> (write all SHA's to a manifest) -> (clone repos you need from the SHA-manifest) -> build

### Freezing the state of distributions
We already use patchlists, so we can continue to do this. One point of improvement though is perhaps we need a way to guarantee that the git-ref in the patchlist has not been force-updated. 

Since a git-ref (branch or PR) can be rebased, once an engineer updates the git-ref, you can't build the old build anymore. We need to store the patches somewhere. Perhaps we can freeze the diff of the git-ref in some location alongside the `requirements-*.in` and the "SHA manifest".

### Nightly CI
We can have the nightly CI freeze all of this metadata (it would also build from this metadata), and make write it somewhere in version control so that we can fork it for releases.

### Metadata location
There a couple of options we have for metadata location; here are two:
1. Within the JAX-Toolbox git repo (have nightly CI commit to `metadata` branch)
    * pros: simple; authoring CI is easier; will not clutter main branch
    * cons: can cause the repo to grow (especially if we store patch diffs here too)
2. In an auxiliary git repo
    * pros: keeps JAX-Toolbox history clean
    * cons: adds complexity to the process since atomicity is more difficult to guarantee; engineers need to be aware of multiple repos

I'm more in favor or (1) due to simplicity, but am open to other options. FWIW with (1), we can merge in (main + metadata) to form a release branch.

#### Installation/Building
It may be obvious, but since some of these repos have more complex build steps; often requiring `apt` packages, or python packages not listed in their `install_requires`, we'll also need one or more `./install-jax-req.sh` to setup or build any wheels we need.

This could potentially live alongside the "SHA manifest"

```[tasklist]
### Tasks
- [ ] Investigate pip-compile @terrykong 
```

### LINKED ISSUE #230: Add sitrep generation for the base container build workflow

### LINKED ISSUE #231: Add sitrep generation for the JAX core container build workflow

### LINKED ISSUE #232: Add sitrep generation for the upstream T5X container build workflow

### LINKED ISSUE #233: Add sitrep generation for the upstream PAX container build workflow

### LINKED ISSUE #328: Dockerfile.jax,pax and t5x try install flax.
The output of some grep:
```
Dockerfile.jax:RUN { pip install flax || true; } && rm -rf ~/.cache/*
Dockerfile.pax.amd64:ADD install-flax.sh /usr/local/bin
Dockerfile.pax.amd64:install-flax.sh --defer
Dockerfile.t5x:ADD install-flax.sh /usr/local/bin
Dockerfile.t5x:install-flax.sh
```

Dockerfile.jax try to install flax, but allows it to fail.
Then t5x and pax Dockerfile install it via install-flax.sh

We should be more consistent. Should we not try to install it in the JAX container?
Or we should use the install-flax.sh file?
This also seem to bypass the pinning of the flax version in the JAX container.

### LINKED ISSUE #337: Fix pip install version conflicts between upper/lower containers

### PR DESCRIPTION
**Pull request by yhtang**
# Summary

- All Python packages, except for a few build dependencies, are now installed using **pip-tools**.
- The JAX and upstream T5X/PAX containers are now built in a two-stage procedure:
  1. The **'meal kit'** stage: source packages are downloaded, wheels built if necessary (for TE, tensorflow-text, lingvo, etc.), but **no** package is installed. Instead, manifest files are created in the `/opt/pip-tools.d` folder to instruct which packages shall be installed by pip-tools. The stage is named due to its similarity in how ingredients in a meal kit are prepared while deferring the final cooking step.
  2. The **'final'** (cooking🔥) stage: this is when pip-tools collectively compile the manifests from the various container layers and then sync-install everything to exactly match the resolved versions.
- Note that downstream containers will **build on top of the meal kit image of its base container**, thus ensuring all packages and dependencies are installed exactly once to avoid conflicts and image bloating.
- The meal kit and final images are published as
  - mealkit: `ghcr.io/nvidia/image:mealkit` and `ghcr.io/nvidia/image:mealkit-YYYY-MM-DD`
  - final: `ghcr.io/nvidia/image:latest` and `ghcr.io/nvidia/image:nightly-YYYY-MM-DD`

# Additional changes to the workflows

- `/opt/jax-source` is renamed to `/opt/jax`. The `-source` suffix is only added to packages that needs compilation, e.g. XLA and TE.
- The CI workflow is now matricized against CPU arch.
- The reusable `_build_*.yaml` workflows are simplified to build only one image for a single architecture at a time. The logic for creating multi-arch images is relocated into the `_publish_container.yaml` workflows and involved during the nightly runs only.
- TE is now built as a wheel and shipped in the JAX core meal kit image.
- TE unit tests will be performed using the upstream-pax image due to the dependency on praxis.
- Build workflows now produce sitreps following the paradigm of #229.
- Removed the various one-off workflows for pinned CUDA/JAX versions.
- Refactored the PAX arm64 Dockerfile in preparation for #338

# What remains to be done

- [ ] Update the Rosetta container build + test process to use the upstream T5X/PAX mealkit (ghcr.io/nvidia/upstream-t5x:mealkit, ghcr.io/nvidia/upstream-pax:mealkit) containers

# Reviewing tips

This PR requires a multitude of reviewers due to its size and scope. I'd truly appreciate code owners to review any changes related to their previous contributions. An incomplete list of reviewer-scope is:
- @terrykong, @ashors1, @sharathts, @maanug-nv: Rosetta, TE, T5X and PAX MGMN tests
- @nouiz: JAX, TE and T5X build
- @joker-eph: PAX arm64 build
- @nluehr: Base image, NCCL, PAX
- @DwarKapex: base/JAX/XLA build, workflow logic

Closes #223
Closes #230 
Closes #231 
Closes #232 
Closes #233 
Closes #271
Fixes #328
Fixes #337 

Co-authored-by: Terry Kong <terryk@nvidia.com>

### CODE REVIEW

**Comment by nouiz** [comment]
What is the replacement for `Removed the various one-off workflows for pinned CUDA/JAX versions.` in the description? Do we still build some multi-version CUDA containers?

**Comment by yhtang** [comment]
> What is the replacement for `Removed the various one-off workflows for pinned CUDA/JAX versions.` in the description? Do we still build some multi-version CUDA containers?

There is no replacement yet. Do we still need these containers? I remember these were built for:
- Selene testing: Selene is getting CUDA 12.2 very soon.
- 23.11 release pinning: it is done, and future release will be handled by dl/dgx/jax

File: .github/container/Dockerfile.base
Comment (terrykong) [inline .github/container/Dockerfile.base:35]: There was some feedback earlier about how this could silently have commits become authored by the jax@nvidia.com user, which caused some issues during CLA presubmits. I reverted it for rosetta here: https://github.com/NVIDIA/JAX-Toolbox/pull/370. What do you think about reverting it here after it's no longer needed?

File: .github/container/Dockerfile.pax.arm64
Comment (terrykong) [inline .github/container/Dockerfile.pax.arm64:78]: The previous stages seem to relate to TF, what is the significance of removing dataclasses as a dep?

File: .github/container/get-source.sh
Comment (terrykong) [inline .github/container/get-source.sh:15]: Looks like --install and --manifest are mutually exclusive; was this intended? Under what scenario would we use --install?

File: .github/container/install-pax.sh
Comment (terrykong) [inline .github/container/install-pax.sh:3]: Should this file be deleted?

File: .github/container/Dockerfile.pax.arm64
Comment (yhtang) [reply .github/container/Dockerfile.pax.arm64:78]: This is inherited from https://github.com/NVIDIA/JAX-Toolbox/blob/4e4e137afd884ebfa748cec636ea751dc8d7ff7a/.github/container/install_lingvo_aarch64.sh#L41. @joker-eph Do you still remember the reason for doing so?

File: .github/container/Dockerfile.base
Comment (yhtang) [reply .github/container/Dockerfile.base:35]: What is the exact CLA issue? Can we use some other username/email that do not annoy the CLA presubmits?

File: .github/container/install-pax.sh
Comment (yhtang) [reply .github/container/install-pax.sh:3]: Yes indeed

File: .github/container/Dockerfile.pax.arm64
Comment (nouiz) [reply .github/container/Dockerfile.pax.arm64:78]: It was needed at some point to install on GH. Maybe that changed, the only way to know is to test it.

**Comment by nouiz** [comment]
This PR is gigantic. If we can't merge it in2-3 days, we should try to split it.
If we get there, we could do it per container.
Make only one PR for the base infra and split the JAX container into 2 steps, meal + final. The other stay based on final. Then one PR for each other containers that start to use the meal container instead of final.

**Comment by yhtang** [comment]
> This PR is gigantic. If we can't merge it in2-3 days, we should try to split it. If we get there, we could do it per container. Make only one PR for the base infra and split the JAX container into 2 steps, meal + final. The other stay based on final. Then one PR for each other containers that start to use the meal container instead of final.

The challenge of splitting would be to rewire the workflows between passing the mealkit and final tags.
Will someone be able to do it, as I'm technically OOTO?

File: .github/container/Dockerfile.base
Comment (terrykong) [reply .github/container/Dockerfile.base:35]: TJ was using our containers to commit code and this global username/email got used to author the commits. When a PR was made, google's CLA bot flagged it as a developer who hadn't signed a CLA.

File: .github/workflows/_publish_container.yaml
Comment (terrykong) [inline .github/workflows/_publish_container.yaml:92]: Does this work instead?

`docker buildx imagetools create $(echo "${{ inputs.SOURCE_IMAGE }}" | tr '\n' ' ') --tag ${{ inputs.TARGET_IMAGE }}`

File: .github/workflows/nightly-te-test.yaml
Comment (terrykong) [inline .github/workflows/nightly-te-test.yaml:15]: Same Q about upstream-pax?
```suggestion
        default: 'ghcr.io/nvidia/upstream-pax:latest'
```

File: .github/workflows/nightly-te-test.yaml
Comment (terrykong) [inline .github/workflows/nightly-te-test.yaml:28]: Same Q about upstream-pax
```suggestion
  DEFAULT_JAX_TE_IMAGE: 'ghcr.io/nvidia/upstream-pax:latest'
```

File: .github/workflows/weekly-base-build.yaml
Comment (terrykong) [inline .github/workflows/weekly-base-build.yaml:54]: Any reason this file doesn't use _publish_container.yaml?

File: .github/container/Dockerfile.base
Comment (terrykong) [reply .github/container/Dockerfile.base:35]: It's my opinion that we should remove the global username/email since this is kind of a silent failure since the end user doesn't get alerted that their commits are authored by jax@nvidia.com

File: .github/container/Dockerfile.pax.arm64
Comment (terrykong) [inline .github/container/Dockerfile.pax.arm64:59]: What do you think about using using a copy heredoc to add the patch to the container and then committing this so it doesn't look like we have unstaged files?

File: .github/workflows/_ci.yaml
Comment (terrykong) [inline .github/workflows/_ci.yaml:34]: I think this should be:
```suggestion
        value: ${{ jobs.build-base.outputs.DOCKER_TAG }}
```

File: .github/container/pip-finalize.sh
Comment (terrykong) [inline .github/container/pip-finalize.sh:7]: This still yields unpinned VCS installs. I believe we still need a mechanism like this: https://github.com/NVIDIA/JAX-Toolbox/pull/271/files#diff-8386cfe5c29a3be212a911c9550a56c57420a17cae190b80f60e52c0ce265645R151-R175

File: .github/workflows/nightly-t5x-build.yaml
Comment (terrykong) [inline .github/workflows/nightly-t5x-build.yaml:68]: Should this be:
```suggestion
      DOCKER_TAG_MEALKIT: ''
      DOCKER_TAG_FINAL: ''
```
?

File: .github/workflows/nightly-te-test.yaml
Comment (yhtang) [reply .github/workflows/nightly-te-test.yaml:6]: Yes indeed. Good catch.

File: .github/workflows/ci.yaml
Comment (terrykong) [inline .github/workflows/ci.yaml:82]: Are you planning on adding a "merge" step here to merge the arm64 and amd64 into a multiarch manifest? Or is the expectation for the concrete workflows `"Nightly ..."`  to provide a "ARCHITECTURE" workflow input?

I'm testing #354 and realizing that if a concrete workflow uses a multiarch image, it will fail using the image from the presubmit, since it's a single arch tag.

File: .github/workflows/nightly-t5x-build.yaml
Comment (terrykong) [inline .github/workflows/nightly-t5x-build.yaml:38]: I see there's new canceling logic. What does this do that the previous implementation didn't?

File: .github/workflows/_publish_container.yaml
Comment (yhtang) [reply .github/workflows/_publish_container.yaml:92]: That won't work because:
1. `inputs.TARGET_IMAGE` is just the part before `:`)
2. `inputs.TARGET_IMAGE` may contain multiple lines of tags using the docker-metadata action format. For example, see https://github.com/NVIDIA/JAX-Toolbox/blob/5ae2e7913558ff134e17c592ea40f0aadc1f04d2/.github/workflows/nightly-jax-build.yaml#L66C12-L66C12

Also, the reason why I use the `manifests` step to pick up the SHAs is to get ride of an 'unknown' platform that gets added by buildx for every image. It is largely cosmetic but could still be confusing for end users.

File: .github/workflows/nightly-te-test.yaml
Comment (yhtang) [reply .github/workflows/nightly-te-test.yaml]: Fixed in 6974a3ad.

File: .github/workflows/nightly-te-test.yaml
Comment (yhtang) [reply .github/workflows/nightly-te-test.yaml]: Fixed in 6974a3ad

File: .github/workflows/weekly-base-build.yaml
Comment (yhtang) [reply .github/workflows/weekly-base-build.yaml:57]: I think I might just forget to to so. Done in ce8ba2d.

File: .github/container/Dockerfile.pax.arm64
Comment (yhtang) [reply .github/container/Dockerfile.pax.arm64:59]: This is for lingvo, which is a transitive dependency built in the `lingvo-builder` stage. The source code is not included in the later mealkit/final stages. Only the wheel is copied over.

File: .github/workflows/_ci.yaml
Comment (yhtang) [reply .github/workflows/_ci.yaml]: Fixed in 6501ae905fad29b4869eba7c7a1f749b088cf616

File: .github/container/pip-finalize.sh
Comment (yhtang) [reply .github/container/pip-finalize.sh:7]: Does this actually pin the versions or only document the SHA of the VCS installs at the time of pip-compile?

File: .github/workflows/ci.yaml
Comment (yhtang) [reply .github/workflows/ci.yaml:82]: Right now we are doing one job per multi-arch image (JAX/PAX/T5X) and it's too verbose. We could add a single job at the end of the presubmit to properly group all single-arch images. It can take advantage of the sitrep jsons to locate the single-arch images.

Comment (yhtang) [reply .github/workflows/ci.yaml:82]: Would it be OK to do it as a follow-up PR?

File: .github/workflows/nightly-t5x-build.yaml
Comment (yhtang) [reply .github/workflows/nightly-t5x-build.yaml:38]: Potentially allowing for more complicated cancellation logic in the metadata job, plus nicer job graph visualization IMHO :)

Comment (yhtang) [reply .github/workflows/nightly-t5x-build.yaml:38]: Previously the logic was limited to what a single-line GitHub Actions expression allows. Now we can do arbitrary bash scripting to determine whether a job needs to be cancelled.

File: .github/container/pip-finalize.sh
Comment (terrykong) [reply .github/container/pip-finalize.sh:7]: IIRC this will pin it

File: .github/workflows/ci.yaml
Comment (terrykong) [reply .github/workflows/ci.yaml:82]: Sure. Let's not block this PR

File: .github/workflows/nightly-t5x-build.yaml
Comment (terrykong) [reply .github/workflows/nightly-t5x-build.yaml:38]: Got it. That's helpful. IIRC, I think there were sometimes race conditions when `if-upstream-failed`'s outcome was "fail", since it wasn't listed in any job's `needs:` section, you could sometimes get farther along on jobs dependent on the `metadata` job before they fail.

File: .github/workflows/nightly-t5x-build.yaml
Comment (terrykong) [reply .github/workflows/nightly-t5x-build.yaml]: This line was updated, but the change was FINAL -> MEALKIT. Should this not have both outputs?

File: .github/workflows/_ci.yaml
Comment (ashors1) [inline .github/workflows/_ci.yaml:203]: Should this be `needs.build-rosetta-t5x.outputs.DOCKER_TAG_FINAL`?

File: .github/workflows/nightly-rosetta-pax-build-test.yaml
Comment (ashors1) [inline .github/workflows/nightly-rosetta-pax-build-test.yaml:131]: `needs.amd64.outputs.DOCKER_TAG_FINAL`?

File: .github/workflows/nightly-rosetta-t5x-build-test.yaml
Comment (ashors1) [inline .github/workflows/nightly-rosetta-t5x-build-test.yaml:132]: I think `ROSETTA_IMAGE` needs to be updated here

File: .github/workflows/nightly-rosetta-t5x-build-test.yaml
Comment (ashors1) [inline .github/workflows/nightly-rosetta-t5x-build-test.yaml:140]: Same as above

File: .github/workflows/_test_pax.yaml
Comment (ashors1) [inline .github/workflows/_test_pax.yaml:152]: I think the singleprocess tests were erroneously removed here. Can we add them back if this is indeed a mistake?

File: .github/workflows/_test_pax_rosetta.yaml
Comment (ashors1) [inline .github/workflows/_test_pax_rosetta.yaml:153]: Same comment about the singleprocess tests

**Comment by ashors1** [comment]
Note that the presubmit CI does not test the `nightly-*.yaml` files. Can we make sure to manually run `nightly-rosetta-{t5x,pax}-build-test.yaml` before merging as well to ensure there are no issues there?

**Comment by yhtang** [comment]
Makes sense. I did test the nightly workflows manually before making the PR, but it's necessary to run them again as the PR has evolved.

File: .github/workflows/_ci.yaml
Comment (DwarKapex) [reply .github/workflows/_ci.yaml:203]: Done

File: .github/workflows/nightly-rosetta-pax-build-test.yaml
Comment (DwarKapex) [reply .github/workflows/nightly-rosetta-pax-build-test.yaml:131]: Agree. But question arises: do we want to run this tests on arm? If yes, it would be better to split this file into 2: build and test. I created a ticket for this: https://github.com/NVIDIA/JAX-Toolbox/issues/376

File: .github/workflows/nightly-rosetta-t5x-build-test.yaml
Comment (DwarKapex) [reply .github/workflows/nightly-rosetta-t5x-build-test.yaml:132]: Done

File: .github/workflows/_test_pax.yaml
Comment (DwarKapex) [reply .github/workflows/_test_pax.yaml:152]: Reverted

File: .github/workflows/_test_pax_rosetta.yaml
Comment (DwarKapex) [reply .github/workflows/_test_pax_rosetta.yaml:153]: Done

**Comment by terrykong** [comment]
FYI @yhtang  and @DwarKapex. Here is a PR rebased on this branch (before Vlad's changes) that added reproducible patches, but now incorporates the mealkit/final mechanic.

**Comment by yhtang** [comment]
> FYI @yhtang and @DwarKapex. Here is a PR rebased on this branch (before Vlad's changes) that added reproducible patches, but now incorporates the mealkit/final mechanic.

Seems like the PR number is missing?

**Comment by terrykong** [comment]
Apologies. Yes, @yhtang #354 is the one

File: .github/container/pip-finalize.sh
Comment (terrykong) [reply .github/container/pip-finalize.sh:7]: After speaking with @DwarKapex , we will just address this in another PR since it's better to get this out ASAP. I will file an issue for this

**Comment by terrykong** [comment]
Created issue to track the VCS installs: https://github.com/NVIDIA/JAX-Toolbox/issues/384 @DwarKapex  and @yhtang

File: .github/container/Dockerfile.base
Comment (yhtang) [reply .github/container/Dockerfile.base:35]: I see. So the container should not ship information that are user-specific. Makes sense.

File: .github/workflows/nightly-rosetta-pax-build.yaml
Comment (terrykong) [inline .github/workflows/nightly-rosetta-pax-build.yaml:14]: It looks like this default won't work anymore. Can the description and default be updated?

This image would result in:
```
BASE_IMAGE_AMD64=ghcr.io/nvidia/upstream-pax:mealkit-amd64-mealkit
BASE_IMAGE_ARM64=ghcr.io/nvidia/upstream-pax:mealkit-arm64-mealkit
```

My opinion is default = '',  and the description should be updated to specify it only accepts jax-toolbox internal images that get single arch tags

File: .github/workflows/nightly-rosetta-t5x-build-test.yaml
Comment (terrykong) [inline .github/workflows/nightly-rosetta-t5x-build-test.yaml:31]: Same comment about inputs.BASE_IMAGE's default/description in rosetta pax's build-test workflow

File: rosetta/Dockerfile.pax
Comment (terrykong) [inline rosetta/Dockerfile.pax:28]: Any reason the `bash -e` was removed?

File: rosetta/Dockerfile.t5x
Comment (terrykong) [inline rosetta/Dockerfile.t5x:28]: Same Q about the `bash -e` being removed

File: .github/workflows/nightly-rosetta-pax-build.yaml
Comment (DwarKapex) [reply .github/workflows/nightly-rosetta-pax-build.yaml:14]: Yeap, agree. Updated the value

File: rosetta/Dockerfile.pax
Comment (DwarKapex) [reply rosetta/Dockerfile.pax:28]: Reverted after PM

File: rosetta/Dockerfile.t5x
Comment (terrykong) [inline rosetta/Dockerfile.t5x:45]: This should be /opt/flax

File: rosetta/Dockerfile.pax
Comment (terrykong) [inline rosetta/Dockerfile.pax:39]: Was moving this from line 38 down intentional?

File: rosetta/Dockerfile.t5x
Comment (terrykong) [inline rosetta/Dockerfile.t5x:34]: Same Q about moving bash -e

**Review by DwarKapex — APPROVED** [review]

**Review by terrykong — APPROVED** [review]

**Pull request merged**
